common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+ * imposes a lightweight read lock, by inserting an entry into the
+ * cache with State.ReadInProgress. A write does the same, with
+ * State.WriteInProgress.
+ *
+ * On read or write completion, the value transitions to
+ * State.Cached.
+ *
+ * State.Initial is represented implicitly by absence from the cache.
+ *
+ * The handshake for cache entry state transition is:
+ * 1. if entry is in an agreeable state, proceed
+ *    where agreeable for reads is Initial, ReadInProgress, or Cached
+ *                    and the read proceeds by ensuring the entry is State.ReadInProgress
+ *      and agreeable for writes is Initial or Cached
+ *                    and the write proceeds by ensuring the entry is State.WriteInProgress
+ *
+ * 2. if entry is not in an agreeable state, then read- or write-around the cache
+ *
+ * 3. to swap in the new state to an existing entry, we use an AtomicReference.compareAndSet
+ *
+ * 4. only if the db operation completes with success, atomically set the state to Cached.
+ *
+ */
+trait MultipleReadersSingleWriterCache[W, Winfo] {
+    /** Subclasses: Toggle this to enable/disable caching for your entity type. */
+    protected def cacheEnabled = true
+
+    /** Subclasses: tell me what key to use for updates */
+    protected def cacheKeyForUpdate(w: W): Any
+
+    /** Each entry has a state, as explained in the class comment above */
+    protected object State extends Enumeration {
+        type State = Value
+        val ReadInProgress, WriteInProgress, InvalidateInProgress, Cached = Value
+    }
+    import State._
+    implicit def state2Atomic(state: State): AtomicReference[State] = {
+        new AtomicReference(state);
+    }
+
+    /** Failure modes, which will only occur if there is a bug in this implementation */
+    case class ConcurrentOperationUnderRead(actualState: State) extends Exception(s"Cache bug: a read started, but completion raced with a concurrent operation: ${actualState}")
+    case class ConcurrentOperationUnderUpdate(actualState: State) extends Exception("Cache bug: an update started, but completion raced with a concurrent operation: ${actualState}")
+
+    /**
+     * The entries in the cache will be a triple of (transid, State, Future[W])
+     *
+     * We need the transid in order to detect whether we have won the race to add an entry to the cache
+     *
+     */
+    private case class Entry(transid: TransactionId, state: AtomicReference[State], value: Future[W]) {
+        def invalidate = state.set(InvalidateInProgress)
+
+        def writeDone()(implicit logger: Logging): Boolean = {
+            logger.debug(this, "Write finished")
+            trySet(WriteInProgress, Cached)
+        }
+
+        def readDone()(implicit logger: Logging): Boolean = {
+            logger.debug(this, "Read finished")
+            trySet(ReadInProgress, Cached)
+        }
+
+        def trySet(expectedState: State, desiredState: State): Boolean = {
```

Comment:
Is this method worthwile? Would it make sense to just call `state.compareAndSet` above?